### PR TITLE
Tech 5112 create declare schema block dsl

### PR DESCRIPTION
**Pull request by dcaddell**
### Deprecated
- The `fields` dsl method is being deprecated.

### Added
- Added the `declare_schema` method to replace `fields`. We now expect a column's type to come before the name
i.e. `declare schema { string :title }`. Otherwise, there is no difference between `fields` and `declare_schema`.

### CODE REVIEW

File: spec/lib/declare_schema/interactive_primary_key_spec.rb [hunk 1]
```diff
@@ -11,49 +11,101 @@
...
+        end
+        self.primary_key = "id"
       end
-      self.primary_key = "id"
-    end
 
-    puts "\n\e[45m Please enter 'id' (no quotes) at the next prompt \e[0m"
-    generate_migrations '-n', '-m'
-    expect(Foo._defined_primary_key).to eq('id')
+      allow_any_instance_of(DeclareSchema::Support::ThorShell).to receive(:ask).with(/one of the rename choices or press enter to keep/) { 'id' }
```
Comment (yinonrousso) [inline spec/lib/declare_schema/interactive_primary_key_spec.rb:33]: Nicely done!

File: lib/declare_schema/dsl.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      field(:created_at, :datetime, null: true)
+      field(:updated_at, :datetime, null: true)
+    end
+
+    def optimistic_lock
+      field(:lock_version, :integer, default: 1, null: false)
+    end
+
+    def field(name, type, *args)
+      options = args.extract_options!
```
Comment (ColinDKelley) [inline lib/declare_schema/dsl.rb:32]: I think we can get rid of `extract_options!` here:
```
def field(name, type, *args, **options)
```

**Review by ColinDKelley — APPROVED** [review]
Looks good! Just one suggestion to remove a line.

File: lib/generators/declare_schema/support/model.rb [hunk 3]
```diff
@@ -69,11 +69,11 @@ def inject_declare_schema_code_into_model_file
             def declare_model_fields_and_associations
               buffer = ::DeclareSchema::Support::IndentedBuffer.new(indent: 2)
               buffer.newline!
-              buffer << 'fields do'
+              buffer << 'declare_schema do'
               buffer.indent! do
                 field_attributes.each do |attribute|
-                  decl = "%-#{max_attribute_length}s" % attribute.name + ' ' +
-                    attribute.type.to_sym.inspect +
+                  decl = "%-#{max_attribute_length}s" % attribute.type + ' ' +
```
Comment (yinonrousso) [inline lib/generators/declare_schema/support/model.rb:75]: Interesting to see how this column and type switch works!

**Review by yinonrousso — APPROVED** [review]
Changes look great!

**Comment by yinonrousso** [comment]
One last minute change is to update the [git workflow](https://github.com/Invoca/declare_schema/blob/master/.github/workflows/declare_schema_build.yml#L60) to remove the `test_responses.txt`

File: .github/workflows/declare_schema_build.yml [hunk 4]
```diff
@@ -57,4 +57,4 @@ jobs:
           git config --global user.email "dummy@example.com"
           git config --global user.name "dummy"
           MYSQL_PORT=3306 bundle exec rake test:prepare_testapp[force]
-          bundle exec rake test:all < test_responses.txt
+          bundle exec rake test:all
```
[the remark this answers is not in the document]
Comment (dcaddell) [reply .github/workflows/declare_schema_build.yml:60]: @yinonrousso is this the change that you were asking for?

[the remark this answers is not in the document]
Comment (yinonrousso) [reply .github/workflows/declare_schema_build.yml:60]: @dcaddell whoops looks like I missed that! Yes it was 👍

**Pull request merged**